doc/proposals/scorecard-custom-tests-2.md

```diff
@@ -0,0 +1,246 @@
+
+## Design Details
+
+### Glossary
+
+ * test image - this is a container image built by end-users to contain a custom test implementation
+ * test configuration (DSL) - this is a file that contains the test definitions and labels which determine what tests are run and how they are configured
+ * Custom Test Definition - A custom test is any sort of test logic required, written in any programming language that satifies the scorecard test requirements for output format and execution.
+
+#### Test Image
+
+A custom test is based on a container image, that image is created by
+the end-user using whatever tools they want to use.
+
+The test image must produce container log output that conforms to the scorecard Test Result output.  The output is an array of Test Results, so a test image could contain multiple tests if necessary.  Here is a sample of test output in the required format:
+```
+    [
+      {
+	      "name": "Spec fields with descriptors",
+	      "description": "All spec fields have matching descriptors in the CSV",
+	      "state": "fail",
+	      "suggestions": [
+		"Add a spec descriptor for size"
+	      ],
+	      "crname": "example-memcached"
+      }, 
+      {
+	      "name": "Some other test",
+	      "description": "Another test on the CSV",
+	      "state": "pass",
+	      "suggestions": [
+	      ],
+	      "crname": "example-memcached"
+      },
+    ]
+```
+
+The test image will be executed as a Pod by the scorecard with a restart policy of `never`.
+
+Also, for each test execution, scorecard would also make the
+bundle image contents available to test images by mounting
+the bundle image contents into the following location inside the test pod:
+
+```
+/scorecard
+```
+
+This allows tests to have access to all the bundle contents if they 
+are required for the test logic.  Note:  It is the responsibility of the test to iterate though each CR that they want to test and create Custom Resources as scorecard will no longer be responsible for this.  The list of CRs to test are found in the CSV within the bundle image content.
+
+For the case of executing a test, the bundle image will be passed as a scorecard command flag:
+```
+operator-sdk scorecard --bundle='quay.io/myorg/mytest' --selector=suite=basic
+```
+
+#### Test Output
+
+Custom tests are expected to produce log output that conforms to the
+scorecard v1alpha2 Test Result output.  This allows scorecard to parse
+the custom test log output and aggregate the results like any other
+scorecard test.
+
+The scorecard expected output would be in JSON, and would match the following definition:  https://github.com/operator-framework/operator-sdk/blob/master/pkg/apis/scorecard/v1alpha2/types.go#L36
+
+The scorecard v1alpha2 Test Result output contains information such as
+pass or fail status and suggestions/errors produced by the test that
+the end user would want to see in the aggregated scorecard results.
+
+
+### Custom Test Development Model
```

Comment:
This feels like a very long list of steps and might not convince a lot of developers to actually ship tests.

How about the following being the default for most users, especially when they start:

1. Developer deploys kuttl-based assertion tests in the form of YAML files in a known directory location
2. SDK has a `test` command which
   - automates the creation of the scorecard config referencing the kuttl tests
   - runs the tests using the scorecard configuration on disk with no particular label selector (all tests)
   - supports a `--namespace` tag, should the Operator prefer a certain watch namespace
3. when the user is happy the SDK has a `build` command which builds the bundle with the test definition and scorecard configuration from 2)

Reply:
@jmccormick2001 @shawn-hurley correct me if I'm wrong but I think the idea is that if an operator developers runs `operator-sdk scorecard` with no options, the following would happen:
1. Scorecard would initialize an empty list of tests 
2.  Scorecard would append the basic test configuration to the list of tests
3. If a bundle directory exists at a known project location, scorecard would append the OLM test configuration to the list of tests
4. If a kuttl directory exists at a known project location, scorecard would append the kuttl test configuration to the list of tests.
5. If a DSL yaml file exists at a known project location, scorecard would append the tests configured in the DSL file to the list of tests.
6. Scorecard runs all of the tests that were setup in steps 2-5.

To package the bundle, the operator would run `operator-sdk bundle create` as usual. We would  add the DSL file and kuttl directory to the bundle if we find them.

I'm thinking we add a `scorecard` directory to the project and store the DSL config file and kuttl directories there. Then the bundle dockerfile just needs `COPY scorecard /scorecard`

Reply:
yep, 
 * we would need the 'operator-sdk bundle create' to be able to do the image build
 * we would run the default basic/olm tests if no selector is added (like we it works today)
 * we need a place to store the scorecard assets (config and other), a 'scorecard' subdir under the project would be fine with me
 * I was thinking that when a person builds a new project it would scaffold out the scorecard subdir with the basic/olm tests being configured by default, I suppose you could also have a 'scorecard --scaffold' command that users could run afterwards to re-scaffold based on what is in the scorecard subdir.

if this works for everyone I could update the proposal to spell it out further.

Reply:
let's talk about the name of the test dir in another followup enhancement specifically in operator-framework/enhancements.

I agree that this is the correct default behavior of the scorecard test.

Reply:
I still have a question about the basic and olm tests. Do we expect to find them explicitly configured in the DSL file, or do we build their configs into the binary and inject them automatically?

What would the scaffolded `scorecard/config.yaml` file look like after `operator-sdk new` (or `init` when we switch to KB's CLI for scaffolding)?

Reply:
If they're explicitly in the file, that gives users the opportunity to permanently remove them without using selectors.

Reply:
I was planning on making them explicit in the config file to allow users to tweak as they want, however, I see the case for just keeping them inside the scorecard by default if we want to force them to always be available.

Reply:
perhaps if the goal with this is to always have a known set of tests (olm/basic) that CVP can assume will always be there, then you'd have to go with the approach of keeping them inside scorecard itself instead of a config file.

Reply:
I would be fine with either method. If we go the explicit route, then the `kuttl` tests should be explicit as well for consistency.

For the pipeline/CI case, what happens if a user can't get the basic tests to pass, so they just remove them from the config file?  Do we envision the pipeline injecting it's own test configs? 

How would that work if the DSL is built into the bundle image and scorecard doesn't give the pipeline a chance to inject anything since it handles the bundle image extraction internally?

Reply:
We might need to answer that question regardless, if there are other policy tests NOT built into scorecard that a pipeline would want to enforce.

Reply:
because they are optional based the fact that a directory exists, I don't think they can be in the config file right?

Reply:
here is what I'm currently thinking:  if a pipeline requires tests to be run, then they would require those tests be added to the config file, if the tests are not in the config file then the pipeline fails the test...kuttl is the only case where the tests that actually are run would be driven by what kuttl tests are found in the kuttl specific directory.  We could start with explicit tests in the config and later address injecting those behind the scenes if it turns out we need to do that for pipeline reasons.